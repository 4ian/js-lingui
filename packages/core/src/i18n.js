/* @flow */
import { interpolate } from "./context"
import { isString, isFunction } from "./essentials"
import { date, number } from "./formats"
import type { DateFormat, NumberFormat } from "./formats"

import * as dev from "./dev"

type MessageOptions = {|
  defaults?: string,
  formats?: Object
|}

type Locale = string
type Locales = Locale | Locale[]

type LocaleData = {
  plurals?: Function
}

type Messages = { [key: string]: string | Function }

type Catalog = {
  messages: Messages,
  languageData: LocaleData
}

type Catalogs = { [key: string]: Catalog }

type setupI18nProps = {
  language?: string,
  locales?: Locales,
  catalogs?: Catalogs,
  missing?: string | Function
}

class I18n {
  _language: string
  _locales: ?Locales

  // Message catalogs
  _catalogs: Catalogs

  _missing: ?string | Function

  _dev: Object

  date: Function
  number: Function

  constructor() {
    // Messages and languageData are merged on load,
    // so we must initialize it manually
    this._catalogs = {}
  }

  get availableLanguages(): Array<string> {
    return Object.keys(this._catalogs)
  }

  get language(): string {
    return this._language
  }

  get locales(): ?Locales {
    return this._locales
  }

  get messages(): Messages {
    return (this._catalogs[this._language] || { messages: {} }).messages
  }

  get languageData(): LocaleData {
    if (process.env.NODE_ENV !== "production") {
      if (
        !this._catalogs[this._language] ||
        !this._catalogs[this._language].languageData
      ) {
        return this._dev.loadLanguageData(this._language)
      }
    }
    return (this._catalogs[this._language] || { languageData: {} }).languageData
  }

  load(locale: Locale, catalog: Catalog) {
    if (typeof catalog !== "object") return

    if (!this._catalogs[locale]) {
      this._catalogs[locale] = {
        messages: {},
        languageData: {}
      }
    }

    Object.assign(this._catalogs[locale].messages, catalog.messages)
    Object.assign(this._catalogs[locale].languageData, catalog.languageData)
  }

  activate(locale: Locale, locales?: Locales) {
    if (!locale) return

    if (process.env.NODE_ENV !== "production") {
      if (this.availableLanguages.indexOf(locale) === -1) {
        console.warn(`Message catalog for locale "${locale}" not loaded.`)
      }
    }

    this._language = locale
    this._locales = locales
  }

  use(language: string, locales?: Locales) {
    const i18n = setupI18n({
      language,
      locales,
      development: this._dev
    })
    i18n._catalogs = this._catalogs
    return i18n
  }

  // default translate method
  _(
    id: string | Object,
    values: Object = {},
    { defaults, formats = {} }: MessageOptions = {}
  ) {
    // Expand message descriptor
    if (id && typeof id === "object") {
      values = id.values
      defaults = id.defaults
      formats = id.formats
      id = id.id
    }

    let translation = this.messages[id] || defaults || id

    // replace missing messages with custom message for debugging
    const missing = this._missing
    if (missing && !this.messages[id]) {
      translation = isFunction(missing) ? missing(this.language, id) : missing
    }

    if (process.env.NODE_ENV !== "production") {
      if (isString(translation) && this._dev && isFunction(this._dev.compile)) {
        translation = this._dev.compile(translation)
      }
    }

    if (!isFunction(translation)) return translation
    return interpolate(
      translation,
      this.language,
      this.locales,
      this.languageData
    )(values, formats)
  }

  pluralForm(
    n: number,
    pluralType?: "cardinal" | "ordinal" = "cardinal"
  ): string {
    if (!this.languageData.plurals) return "other"
    return this.languageData.plurals(n, pluralType === "ordinal")
  }

  date(value: string | Date, format: DateFormat): string {
    return date(this.locales || this.language, format)(value)
  }

  number(value: number, format: NumberFormat): string {
    return number(this.locales || this.language, format)(value)
  }
}

function setupI18n(params?: setupI18nProps = {}): I18n {
  const i18n = new I18n()

  if (process.env.NODE_ENV !== "production") {
    i18n._dev = dev
  }

  if (params.catalogs) {
    Object.keys(params.catalogs).forEach(locale =>
      i18n.load(locale, params.catalogs[locale])
    )
  }
  if (params.language) i18n.activate(params.language, params.locales)
  if (params.missing) i18n._missing = params.missing

  return i18n
}

export { setupI18n }
export type { MessageOptions, Catalog, Catalogs, LocaleData, I18n, Locales }
