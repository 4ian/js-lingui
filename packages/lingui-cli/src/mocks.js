export function mockConsole (testCase, mock = {}) {
  const originalConsole = global.console

  const defaults = {
    log: jest.fn(),
    warn: jest.fn()
  }

  global.console = {
    ...defaults,
    ...mock
  }

  try {
    testCase(global.console)
  } catch (e) {
    global.console = originalConsole
    throw e
  }

  global.console = originalConsole
}
