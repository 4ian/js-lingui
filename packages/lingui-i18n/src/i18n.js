/* @flow */
import MessageFormat from 'messageformat'
import t from './t'
import { select, plural, selectOrdinal } from './select'
import plurals from './plurals'

type Catalog = {[key: string]: string}
type Catalogs = {[key: string]: Catalog}

type Message = {|
  id: string,
  defaults?: string,
  params?: Object
|}

class I18n {
  _language: string
  _messages: Catalogs

  t: Function
  plural: Function
  select: Function
  selectOrdinal: Function

  constructor (language: string = '', messages: Catalogs = {}) {
    this._language = language
    this._messages = messages

    this.t = t(this)
    this.plural = plural(this)
    this.select = select
    this.selectOrdinal = selectOrdinal(this)
  }

  get messages (): Catalog {
    return this._messages[this._language] || {}
  }

  get language (): string {
    return this._language
  }

  load (messages: Catalogs) {
    if (!messages) return

    // deeply merge Catalogs
    Object.keys({ ...this._messages, ...messages }).forEach(language => {
      if (!this._messages[language]) this._messages[language] = {}

      Object.assign(
        this._messages[language],
        messages[language] || {}
      )
    })
  }

  activate (language: string) {
    if (language) this._language = language
  }

  use (language: string) {
    return new I18n(language, this._messages)
  }

  translate ({ id, defaults, params = {} }: Message) {
    const translation = this.messages[id] || defaults || id
    return this.compile(translation)(params)
  }

  compile (message: string) {
    return new MessageFormat(this.language).compile(message)
  }

  pluralForm (n: number, cardinal?: 'cardinal' | 'ordinal' = 'cardinal'): string {
    const forms = plurals[this._language]
    const form = forms[cardinal]
    return form ? form(n) : 'other'
  }
}

export default new I18n()
export { I18n }
export type { Message, Catalog, Catalogs }
