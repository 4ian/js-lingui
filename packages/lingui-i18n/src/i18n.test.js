/* @flow */
import i18n, { setupI18n } from '.'
import { mockConsole, mockEnv } from './mocks'
import linguiDev from './dev'

describe('I18n', function () {
  it('should export default I18n instance', function () {
    expect(i18n).toBeDefined()
  })

  it('should be initialized with empty catalog', function () {
    const i18n = setupI18n()
    expect(i18n.messages).toEqual({})
  })

  it('should bound t method', function () {
    const i18n = setupI18n()
    expect(i18n.t).toBeInstanceOf(Function)

    expect(i18n.t`Message`).toEqual('Message')

    const name = 'Fred'
    expect(i18n.t`Hello ${name}`).toEqual('Hello Fred')
  })

  it('.load should load catalog and merge with existing', function () {
    const messages = {
      en: {
        'Hello': 'Hello'
      }
    }

    const i18n = setupI18n({
      development: linguiDev
    })
    expect(i18n.messages).toEqual({})

    i18n.load({ en: messages.en })
    i18n.activate('en')
    expect(i18n.messages).toEqual(messages.en)

    // fr catalog shouldn't affect the english one
    i18n.load({ fr: { 'Hello': 'Salut' } })
    expect(i18n.messages).toEqual(messages.en)

    i18n.load({ en: { 'Goodbye': 'Goodbye' } })
    // $FlowIgnore: testing edge case
    i18n.load() // should do nothing
    expect(i18n.messages).toEqual({
      'Hello': 'Hello',
      'Goodbye': 'Goodbye'
    })
  })

  it('.loadLanguageData should load language data and merge with existing', function () {
    const languageData = {
      en: {
        plurals: jest.fn(),
        code: 'en-US'
      }
    }

    const i18n = setupI18n({
      language: 'en',
      languageData,
      messages: {en: {}, fr: {}},
    })

    expect(i18n.languageData).toEqual(languageData.en)

    // fr catalog shouldn't affect the english one
    i18n.loadLanguageData({ fr: { plurals: jest.fn(), code: 'fr-FR' } })
    // $FlowIgnore: testing edge case
    i18n.loadLanguageData() // should do nothing
    expect(i18n.languageData).toEqual(languageData.en)
  })

  it('.activate should switch active language', function () {
    const messages = {
      'Hello': 'Salut'
    }

    const i18n = setupI18n({
      language: 'en',
      messages: { fr: messages, en: {} },
      development: linguiDev
    })

    expect(i18n.language).toEqual('en')
    expect(i18n.messages).toEqual({})

    i18n.activate('fr')
    // $FlowIgnore: testing edge case
    i18n.activate() // should do nothing
    expect(i18n.language).toEqual('fr')
    expect(i18n.messages).toEqual(messages)
  })

  it('.activate should throw an error about incorrect language', function () {
    const i18n = setupI18n()

    mockConsole(console => {
      i18n.activate('xyz')
      expect(console.warn).toBeCalledWith(expect.stringContaining('Unknown local'))
    })

    mockConsole(console => {
      mockEnv('production', () => {
        i18n.activate('xyz')
        expect(console.warn).not.toBeCalled()
      })
    })
  })

  it('.use should return new i18n object with switched language', function () {
    const messages = {
      en: {
        'Hello': 'Hello'
      },
      fr: {
        'Hello': 'Salut'
      }
    }

    const i18n = setupI18n({
      language: 'en',
      messages: messages,
      development: linguiDev
    })

    expect(i18n._({ id: 'Hello' })).toEqual('Hello')

    // change language locally
    expect(i18n.use('fr')._({ id: 'Hello' })).toEqual('Salut')

    // global language hasn't changed
    expect(i18n._({ id: 'Hello' })).toEqual('Hello')
  })

  it('._ should format message from catalog', function () {
    const messages = {
      'Hello': 'Salut',
      'My name is {name}': 'Je m\'appelle {name}'
    }

    const i18n = setupI18n({
      language: 'fr',
      messages: { fr: messages },
      development: linguiDev
    })

    expect(i18n._({ id: 'Hello' })).toEqual('Salut')
    expect(i18n._({ id: 'My name is {name}', values: { name: 'Fred' } }))
      .toEqual("Je m'appelle Fred")

    // missing { name }
    expect(i18n._({ id: 'My name is {name}' }))
      .toEqual("Je m'appelle")

    // Untranslated message
    expect(i18n._({ id: 'Missing message' })).toEqual('Missing message')
  })
})
