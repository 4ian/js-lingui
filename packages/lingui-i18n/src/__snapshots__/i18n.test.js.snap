exports[`I18n .activate should throw an error about incorrect language 1`] = `"Unknown locale \"xyz\"."`;

exports[`I18n .activate should throw an error about incorrect language 2`] = `"Unknown locale \"xyz\"."`;
